### ISSUE TITLE
Specify how Logs SDK implements Enabled

### ISSUE THREAD

**Issue by pellared** [open-telemetry/opentelemetry-specification#4207]
Context: https://github.com/open-telemetry/opentelemetry-specification/pull/4203#issuecomment-2341410062

Blocked by:
- https://github.com/open-telemetry/opentelemetry-specification/pull/4203

**Comment by pellared** [comment]
The idea is to add `OnEnabled` operation to the `LogRecordProcessor`.

The parameters of `OnEnabled` would be defined in the same way as `Logger.Enabled`. However, it is important to notice that in future the parameters may independently change (for instance, the `LogRecordProcessor.OnEnabled` may start accepting an instrumentation scope to allow dynamic filtering of loggers by name). Still, I propose to start with the minimal set of parameters.

The return value of `OnEnabled` would be defined in the same way as `Logger.Enabled`. 

When the user calls `Logger.Enabled`, the SDK would call `OnEnabled` in the same order as processors were registered until any of the processors' returns `true`. `Logger.Enabled` returns false if none of registered processors returns `true`. Pseudocode:

```go
func (l *logger) Enabled(ctx context.Context, param log.EnabledParameters) bool {
	newParam := l.newEnabledParameters(ctx, param)
	for _, p := range l.provider.processors {
		if enabled := p.Enabled(ctx, newParam); enabled {
			return true
		}
	}
	return false
}
```

Pre-work: https://github.com/open-telemetry/opentelemetry-go/pull/5816

**Comment by pellared** [comment]
**Proposal B**

Based on following spike: https://github.com/open-telemetry/opentelemetry-go/pull/5825

The idea is to add a new abstraction named `Filterer` (which is inspired by sampler in tracing SDK) with a `Filter` method.

The parameters of `Filter` would be defined in the same way as `Logger.Enabled`. However, it is important to notice that in future the parameters may independently change (for instance, the `Filterer.Filter` may start accepting an instrumentation scope to allow dynamic filtering of loggers by name). Still, I propose to start with the minimal set of parameters.

The return value of `Filter` would be defined in the same way as `Logger.Enabled`. 

Pseudocode:

```go
type Filterer interface {
	Filter(ctx context.Context, param FilterParameters) bool
}
```

When the user calls `Logger.Enabled`, the SDK would call `Filter` in the same order as filterers were registered until any of the processors' returns `false` (then it breaks and returns `false`. `Logger.Enabled` returns `true` if none of registered filterers returns `true`. Pseudocode:

```go
// Enabled returns false if at least one Filterer held by the LoggerProvider
// that created the logger will return false for the provided context and param.
func (l *logger) Enabled(ctx context.Context, param log.EnabledParameters) bool {
	newParam := l.newEnabledParameters(param)

	for _, flt := range l.provider.filterers {
		if !flt.Filter(ctx, newParam) {
			return false
		}
	}
	return true
}
```

When the user calls `Logger.Emit`, the SDK would first check if the record should be filtered out. Pseudocode:

```go
func (l *logger) Emit(ctx context.Context, r log.Record) {
	param:= l.toEnabledParameters(r)
	if !l.Enabled(ctx, param) {
		return
	}

	newRecord := l.newRecord(ctx, r)
	for _, p := range l.provider.processors {
		if err := p.OnEmit(ctx, &newRecord); err != nil {
			otel.Handle(err)
		}
	}
}

```

Pseudocode of a custom minimum severity level filterer :

```go
type MinSevFilterer struct {
	Minimum api.Severity
}

func (f *MinSevFilterer) Filter(ctx context.Context, param log.FilterParameters) bool {
	lvl, ok := param.Severity()
	if !ok {
		return true
	}
	return lvl >= p.Minimum
}
```

This proposal assumes that each logger provider is used to define a separate processing pipeline. It is more aligned with the current specification design for all signals.

Filtering is coupled to emitting log records. Therefore, a custom processor filtering implementation does not need to implement filtering on both `OnEmit` and `OnEnabled` as in Proposal A. It makes also makes defining filters easier for users.

At last such design should be easier to be added to existing SDKs as it adds a new abstraction instead of adding more responsibilities to `Processor`.

**Comment by MSNev** [comment]
I'm a little concerned that by "defining" the way that "enabled" should be implemented (especially adding layers of filtering (as part of the required implementation)) would (potentially) worse from a perf perspective than just constructing and emitting the Log that eventually gets dropped. ie. the whole point of the "Is enabled" is for avoid perf impact...

**Comment by MSNev** [comment]
I think it's fine to define "what" options should be used (ie the parameters) to "check" whether something is enabled or not and this should be limited.
If an end-user application wants to do something more complicated then the "better" option would be to have "them" (not OTel) implement their own "Filtered" logger thingy.

**Comment by pellared** [comment]
> I'm a little concerned that by "defining" the way that "enabled" should be implemented (especially adding layers of filtering (as part of the required implementation)) would (potentially) worse from a perf perspective than just constructing and emitting the Log that eventually gets dropped. ie. the whole point of the "Is enabled" is for avoid perf impact...

I am not sure if I follow.

`Logger.Enabled` gives the possibility to for the user to check if it is even necessary to construct and emit a big and expensive log record.

Which part of the implementation proposal (A or B or both) are you concerned about?

**Comment by pellared** [comment]
> I think it's fine to define "what" options should be used (ie the parameters) to "check" whether something is enabled or not and this should be limited.
> If an end-user application wants to do something more complicated then the "better" option would be to have "them" (not OTel) implement their own "Filtered" logger thingy.

Agree. PTAL https://github.com/open-telemetry/opentelemetry-specification/pull/4203 where I try to propose a minimal set of parameters.

**Comment by MSNev** [comment]
> I am not sure if I follow.
> 
> `Logger.Enabled` gives the possibility to for the user to check if it is even necessary to construct and emit a big and expensive log record.
> 
> Which part of the implementation proposal (A or B or both) are you concerned about?

Prescribing that it must be done via a specific method (A or B), rather than letting the language / environment determine the best way to provide the capability.

**Comment by MrAlias** [comment]
@MSNev the issue with not specifying behavior is that it can be specified later in a non-compatible way. Meaning, if an implementation implements A and the spec later specifies B that language SIG is non-compliant. Not having some language about implementation both stops implementations from adopting this feature for that reason and causes specification maintainer to not approve any further changes[^1].

This issue it trying to break out of this circular logic.

If you think things should be left undefined, please work with @jack-berg to reach a consensus as his statements ask for the opposite.

[^1]: https://github.com/open-telemetry/opentelemetry-specification/pull/4203#pullrequestreview-2292967830

**Comment by jack-berg** [comment]
What about adding a severity level as part of the [logger config](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#loggerconfig)?

It seems to match the common log workflow of changing severity threshold for specific _loggers_.

I'm suspicious of the idea of introducing the concept of an expandable set of `EnabledParameters`, since requiring users to create a new struct / object to check if code should proceed to emitting a log seems non-optimal from a performance and ergonomics standpoint. I understand the desire to want to retain flexibility to evolve the API, but I think getting the arguments right in the initial stable version and not trying to retain the ability to evolve is likely to produce the most ergonomic API.

**Comment by pellared** [comment]
> What about adding a severity level as part of the [logger config](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#loggerconfig)?

I think that it is reasonable to improve the logging configuration than introducing a new concept.
It also does not mean that a feature like "filterers" cannot be added in future to allow more flexibility/customization if there would be a bigger community demand.

> It seems to match the common log workflow of changing severity threshold for specific loggers.

I agree. I assume that setting a global severity threshold is also a common workflow.

> I'm suspicious of the idea of introducing the concept of an expandable set of EnabledParameters, since requiring users to create a new struct / object to check if code should proceed to emitting a log seems non-optimal from a performance and ergonomics standpoint. 

If you are mentioning EnabledParameters for `Logger.Enabled`? If so then I disagree. The caller is NOT required to call `Logger.Enabled`. If the user would not call it the `Logger.Emit` would still filter out the log records. `Logger.Enabled` gives  the caller the possibility to know whether it make sense to even emit a log. This is important in places when the constructed log record is "heavy" (e.g. lots of attributes) and it is desired to avoid constructing and emitting log records to increase the performance.

**Comment by XSAM** [comment]
Per discussion with @pellared, if the logger can have an `Enabled` method, we probably don't need an `Enabled` method for the processor, as the logger can solve the memory allocation optimization at the beginning of the logging workflow.

**Comment by MrAlias** [comment]
> What about adding a severity level as part of the [logger config](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#loggerconfig)?
> 
> It seems to match the common log workflow of changing severity threshold for specific _loggers_.
> 
> I'm suspicious of the idea of introducing the concept of an expandable set of `EnabledParameters`, since requiring users to create a new struct / object to check if code should proceed to emitting a log seems non-optimal from a performance and ergonomics standpoint. I understand the desire to want to retain flexibility to evolve the API, but I think getting the arguments right in the initial stable version and not trying to retain the ability to evolve is likely to produce the most ergonomic API.

If we unify on the logger config, we would not be able to have different severity level support for different processors. If you have one processor that wants to send to short-term storage and hold all debug logs and a long-term storage that only store error logs and above.

**Comment by pellared** [comment]
> If we unify on the logger config, we would not be able to have different severity level support for different processors. [...]

I think that **Proposal A** gives the biggest amount of flexibility in terms of multiple processing pipelines.

**Comment by pellared** [comment]
Wouldn't it be acceptable if the SDK supports to multiple ways to implement Enabled? E.g. [**Proposal A**](https://github.com/open-telemetry/opentelemetry-specification/issues/4207#issuecomment-2352291363) and [**Proposal C**](https://github.com/open-telemetry/opentelemetry-specification/issues/4207#issuecomment-2362290968).

I think that an ability to configure a logger to make it disabled or specify a minimum severity level would fulfill most use cases and may be well performing. While composing processors may be necessary to allow setting up complex processing pipelines.

**Comment by lmolkova** [comment]
> The possible workaround would be to compose logger providers as noted in proposal B. However, I do not feel like such workaround is good/acceptable.

FWIW that's what .NET ILogger does - [if any of the destinations have the logger + severity enabled, the logging is enabled](https://github.com/dotnet/runtime/blob/b9ffdb0b1b984bc0b06ad51242ef8e14b08e8e6b/src/libraries/Microsoft.Extensions.Logging/src/Logger.cs#L66). 

Since every provider owns it's own configuration via [LoggerConfigurator](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#loggerconfigurator), the composite provider would be responsible for merging them and creating the final `LoggerConfig`. 
It also means that filters would be an (optional) feature based on the LoggerConfig. Ootherwise filtering processors and config may start to contradict each other and it'd be best if we could avoid it.

**Comment by pellared** [comment]
@open-telemetry/technical-committee, can I propose a PR based on Proposal A and Go implementation?

**Comment by tigrannajaryan** [comment]
[Meta] I am commenting to help with the process, not to express an opinion about the proposed approaches.

> @open-telemetry/technical-committee, can I propose a PR based on Proposal A and Go implementation?

@pellared  We definitely need a PR with _a proposal_ and then if it gets approved we need to merge that PR to the spec in _Development_ status first. To emphasize again, I don't have an opinion about which of the proposed approaches we should go ahead with (I think I can see 3 different proposals in this thread).

**Comment by tigrannajaryan** [comment]
>FYI this is blocking Go stability of the Logs signal.

@MrAlias I think based on the agreed process we have to do this:
1. Create a spec PR with a proposal about how this capability is supposed to work. (what @pellared wants to do). Merge it in _Development_ status.
2. Implement the capability in 3 languages in _Development_ status. We have it in Go, which is good, we need 2 more languages (if we don't have them already).
3. Publish implementations, collect community feedback (for a few weeks at least?).
4. If we are happy with implementations mark the capability Stable in the spec.
5. Mark the capability in language implementations Stable. We can also iterate Steps 4 and 5 with in-between maturity levels if we don't want to jump from Development to Stable in one leap.

I think you are saying this issue is blocking Step 5, which of course is true, I just want to make it clear there are a few additional steps before we get there. If you are expecting something else to happen then I may be misunderstanding what you are asking for.

P.S. The requirement for 3 language implementations was discussed but I still don't see it anywhere in the spec. We need to add it.

**Comment by MrAlias** [comment]
Please also see C++ and Rust

**Comment by MrAlias** [comment]
Yeah, not expecting a direct jump. Just looking for continued progress here.

**Comment by MrAlias** [comment]
There have been TC members taking an active role in blocking progress here:

- https://github.com/open-telemetry/opentelemetry-specification/pull/4262
- https://github.com/open-telemetry/opentelemetry-specification/issues/4256

I would appreciate a more active role from the TC in progressing this initiative. Is it possible to add @pellared to the TC?

**Comment by pellared** [comment]
@tigrannajaryan, I am not working on PR as the label is `triage:deciding` and not `triage:accepted`. I do not want to waste the time of the reviewers and contributors until it is clear agreement that we can work on it.

**Comment by jack-berg** [comment]
> There have been TC members taking an active role in blocking progress here:

I've been blocking progress on the grounds of an incomplete proposal. I'll happily concede (though I'll still disagree) if the consensus is that we can / should evolve the API features independently of how those will be used in the SDK.

> I would appreciate a more active role from the TC in progressing this initiative.

Here's my take:

While I understand why you might want the Logger, Meter, Tracer enabled methods to accept `Context`, I think including it as a parameter is **not a good design**.

Let's take stock of some of the API / SDK facilities we have which are related to this problem:

- Tracing SDK has Sampler, which accepts context as an argument, and which is invoked when starting a span.
- Tracing API has Tracer#enabled, which accepts no arguments, and which is used by a caller to avoid even starting a span.
- Tracing SDK has TracerConfigurator, which computes the TracerConfig for a Tracer. TracerConfig has a boolean `disabled` parameter, which is used to determine the response to Tracer#enabled.
- Metrics API instruments have Instrument#enabled, which accepts no arguments, and which is used by a caller to avoid recording measurements.
- Metrics SDK has MeterConfigurator, which computes the MeterConfig for a Meter. MeterConfig has a boolean `disabled` parameter, which is used to determine the response to Instrument#enabled for all instruments registered under the meter.
- Logs API has Logger#enabled, which accepts no arguments, and which is used by a caller to avoid emitting a log.
- Logs SDK has LoggerConfigurator, which computes the LoggerConfig for a Logger. LoggerConfig has a boolean `disabled` parameter, which is used to determine the response to Logger#enabled.
- Log SDK **notably** does not have a Sampler concept, although the idea has been floated.

If we extend Logger#enabled, Instrument#enabled to accept Context, then it follows that we should do the same for Tracer#enabled for the sake of symmetry.

Adding a Context argument to these #enabled methods suggests TracerConfig, MeterConfig, LoggerConfig would be evolved to have some sort of feature where users can provide a function which accepts Context (and maybe other parameters) and return a boolean indicating whether the Logger, Instrument, Tracer is enabled. If the SDK corollary to the API isn't TracerConfig, MeterConfig, LoggerConfig, we'll have to make up some new concept, which I think we should try to avoid.

Adding a programmable extension point to TracerConfig, MeterConfig, LoggerConfig is a sloppy design. Users now have to choose between writing code which turns things on and off in response to context in _either_ a sampler, or in the {Signal}Config. When is it appropriate to use one or another?

A cleaner design is to:

- Keep TracerConfig, MeterConfig, LoggerConfig static as a design principle.
- Limit user-provided implementations of this kind of logic to samplers.
- Provide a new log SDK sampler analagous to trace SDK sampler. 

The benefit of keeping TracerConfig, MeterConfig, LoggerConfig static is that static config lends itself to [declarative configuration](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/configuration#declarative-configuration), where as breaking this principle and accepting user-provided functions encourages bespoke, programmatic configuration. While more flexible, writing custom functions is a worse user experience. If we can carve out a portion of configuration which is static and rule-based, the ecosystem will have better user ergonomics and language interoperability in the long run.

As for what a missing log sampler might look like, we need it to have some symmetry to trace SDK sampler, but also reflect the differences between logging and tracing. In particular,  spans have a start and end stage, and trace Sampler can only receive the bits of the span available at span start. In contrast, all bits of a log record are known at a time when it is omitted. Reflecting these differences, log sampler's might consist of:

- A `shouldSample` operation accepting Context, Severity, Body, and Attributes and returning a boolean. Omit reosurce, scope, timestamp, observed timestamp for consistency with tracing. Omit severityText because it mostly duplicates severity. Return a boolean instead of `SamplingResult` returned by tracing because log sampling is less complicated than trace sampling.
- Call `shouldSample` for each log emitted, and skip calling `LogRecordProcessor#onEmit` if sampler returns false.
- Have build in log samplers for common scenarios:
  - always_on: always samples all log records
  - always_off: always drops all log record
  - trace_based: if there is a span in the context, sample log records where the span is sampled

One thing that's missing is the equivalent programmable extension point for metrics. I would solve this by introducing something like what's been called a measurement processor. An extension point that allows users to intercept all measurements before they're aggregated and do things like: transform value (i.e. unit conversions), drop, enrich with attributes from context, baggage. If this type of capability existed, a measurement processor could make decisions on the basis of Context, which would surely be an argument provided as part of the measurement.

Taking a step back, the proposal is to add a Context argument to the Logger#enabled, Instrument#enabled, Tracer#enabled methods. Since we make no guarantee that these methods are called by instrumentation, the only purpose they can possibly serve is to improve performance. By adding Context as an argument, we're implying some corresponding programmable SDK capability to respond based on the contents of Context. But its **antithetical** to call a user-provided function on an operation that is supposed to be blazing fast and eliminate overhead. 

Let's do the sensible thing and establish a design principle where Tracer#enabled, Instrument#enabled, and Logger#enabled are _always_ statically resolved.

**Comment by pellared** [comment]
I am not sure what "clean' design means here but I think that the config design misses the fundamental functionality of having different log record processing pipelines. E.g. https://github.com/open-telemetry/opentelemetry-specification/issues/4207#issuecomment-2377517345 or different processing for events and bridged log records.

> If we can carve out a portion of configuration which is static and rule-based, the ecosystem will have better user ergonomics

I find this statement as opinionated. Static configuration is less flexible and will be limiting the users.

> But its antithetical to call a user-provided function on an operation that is supposed to be blazing fast and eliminate overhead.

I find this statement as not correct. Does it mean that we do not allow any customization for something that is supposed to be fast? Does it mean that the design of https://pkg.go.dev/log/slog or https://docs.rs/log/latest/log/ is antithetical?

**Comment by jack-berg** [comment]
> I find this statement as opinionated. Static configuration is less flexible and will be limiting the users.

Yes. I'm taking an opinionated stance on what I think the design should be. Taking away flexibility (or put another way, adding guardrails) can be of great benefit to the user experience. After all, the removal of flexibility is what gives programming languages increased productivity and safety. 

> I never proposed a dynamically configured loggerconfig. My preferred proposal is to extend LogRecordProcessor with OnEnabled. 

Still results in the ability for user code to run on the hot path. And the function has to run twice, right? Once when `Logger#enabled` is called, and a second time when the SDK is deciding whether to pass the log to a processor. The second time being necessary because instrumentation isn't required to call `Logger#enabled`. So I believe that's two calls for each registered LogRecordProcessor which provides a filter function, and also the need to iterate through a list of processors. I prefer a design where Logger#enabled is resolved by reading a boolean in memory - that's it.

To be honest, the design to prioritize multiple independent logging pipelines, each with their own filter, doesn't resonate with me. Its too different from the other signals, and adds complexity for what I consider to be a niche use case. I think those types of use cases are better served by the collector, instead of adding the burden to all the language implementations.

> I find this statement as not correct. Does it mean that we do not allow any customization for something that is supposed to be fast? Does it mean that the design of https://pkg.go.dev/log/slog or https://docs.rs/log/latest/log/ is antithetical?

Presumably there's parts of those systems which encourage the pattern of calling user code in a similar scenario as our Logger#enabled operation? If so, I'm not familiar with those systems and can't comment. 

This is all my personal opinion. If there's a consensus among community members to go in a different direction, the community should take that direction.

**Comment by pellared** [comment]
> Presumably there's parts of those systems which encourage the pattern of calling user code in a similar scenario as our Logger#enabled operation? 

This is correct. These systems are core logging abstractions for Go and Rust. A similar pattern is in .NET. See [here](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger?view=net-8.0). Allowing the have a custom hook on `Enabled` is a fundamental feature in these systems.

Therefore, for proposal A we have implementations in 3 languages (with little differences in details), but also a precedence of a similar design in core logging libraries in different languages.

**Comment by pellared** [comment]
Summary from https://cloud-native.slack.com/archives/C062HUREGUV/p1731580335908619:

I start to lean towards **Proposal C - Extend Logger config**  with `min_severity_level` (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#loggerconfig).
I does not require adding any new concept the SDK and should be able to handle the most popular use cases.
It should also be the most performant from the "hot-path" perspective.
I will work on prototyping it in Go.

**Comment by pellared** [comment]
**Proposal A - Extend Processor** looks to be the best solution for users who would like to export using Linux user_events. It needs to by handled dynamically. See: https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/1cb39edbb6467375f71f5dab25ccbc49ac9bf1d5/opentelemetry-user-events-logs/src/logs/exporter.rs#L158-L160

CC @cijothomas

**Comment by pellared** [comment]
Logs SIG meeting notes:
@lmolkova, proposed that the **Proposal C - Extend Config**  could enhanced so that the config does support also dynamic evaluation instead of supporting only declarative configuration.

**Comment by pellared** [comment]
This PR partially solves the issue:
- https://github.com/open-telemetry/opentelemetry-specification/pull/4381

We should still define how the SDK handles `SeverityNumber` passed to `Logger.Enabled`.
Personally, I will work towards:
- https://github.com/open-telemetry/opentelemetry-specification/issues/4363

I also plan to make some intermediate steps like describing how log record filtering and fan-out (multiple isolated pipelines) can be done with the current log processors design. This will basically document the patterns that our users are already using like Decorator and Composite (because there is no other clean way to do it).

**Comment by pellared** [comment]
Closing per https://github.com/open-telemetry/opentelemetry-specification/pull/4381

https://github.com/open-telemetry/opentelemetry-specification/issues/4363 will be towards https://github.com/open-telemetry/opentelemetry-specification/issues/4208
